futurewei-cloud/arktos
# admission controller: annotates pod network-readiness=false by default

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This custom admission controller, "DefaultPodNetworkReadiness", adds annotation "arktos.futurewei.com/network-readiness": "false" to pods, except for those of host network, or already have the annotation (regardless of the value).

External network provider may need to do some network resource preparations on its custom network controller before pod is scheduled and able to start. pod annotation of "arktos.futurewei.com/network-readiness" is introduced as signal to sync up scheduler with the network preparation task before scheduling is made for the pod.

This admission controller is typically required by Arktos working with external VPC network providers. For flat typed network, as there is no controller conducting such network preparations, this admission should not be enabled.

This is part of Arktos support for multi-tenancy multi-network.

**How to verify in local dev env**
1. ensure kube-apiserver will start with this admission controller enabled (in hack/lib/common-var-init.sh)
```ENABLE_ADMISSION_PLUGINS=${ENABLE_ADMISSION_PLUGINS:-"...,DefaultPodNetworkReadiness"}```
2. run hack/arktos-up.sh
3. create a regular pod
4. look at the pod detail spec in yaml format, ensure below annotation is there
```yaml
kind: Pod
metadata:
  annotations:
    arktos.futurewei.com/network-readiness: "false"
```

**Does this PR introduce a user-facing change?**:
YES
```release-note
For Arktos working with VPC network providers like Mizar, Neutron etc, kube-apiserver should start with DefaultPodNetworkNotReady admission controller enabled, via parameter like
    --enable-admission-plugins=...,DefaultPodNetworkReadiness
```
